Strategy pattern collision handler

hungnguyen94 opened the pull request:
- Created Collidable interface, which all (collidable) models should implement. You should use this instead of Shape. 
- Created CollisionAction interface. This interface has a method onCollision(). Every collidable should have a (Hash)Map where different instances of CollisionAction are stored. Depending on the class, a different method will be executed. The use of a Map with the class as key, and an instance of CollisionAction as value, eliminates the need to the instanceof checking.  
- All collision actions in the CollisionHandler are moved to their respective classes. 
- All / most Shape references are refactored to use Collidable.
- Every model also implements the Drawable interface. Every drawable should have a draw method. 
- The Game, Level class can be ignored for now.

Commit:
Start on level class to replace model class

< organisation

Commit:
Add interface Observer and Subject.

Level implements subject.

< organisation, interface

Commit:
Small changes to level class.

< organisation

ccberg committed:
added uml < documentation

ccberg committed:
added uml for levels structure < organisation

ccberg committed:
Merge branch 'uml_levels' into Levels

Commit:
Add game class to start game. Game can be headless

Small changes to observer interface.
Games are started headless in a separate thread. Still needs to connect
with the views.
< interface, organisation

Commit:
Add collision checking in level moveObjects

< interface, organisation

Commit:
Merge branch 'Levels' of github.com:hungnguyen94/BTrouble into Levels

Commit:
Add drawable interface for objects that draw

< interface, organisation

Commit:
Add test example using a view with game

< user-interface

Commit:
Add interfaces for collidables (strategy pattern)

< interface

Commit:
Modifications to interfaces related to collisions

< interface

Commit:
All models implement collidable interface.

< interface

Commit:
Refactored model classes to use Collidable

instead of Shape.

< Interface

Commit:
Collisionhandler uses the Collidable interface

No more instanceof checking, actions are in the collidable classes.

< interface, organisation

Commit:
Refactored test to use Collidible interface

instead of Shape class.

< interface, testing

Commit:
Fix CollisionHandlerTest

< testing

Commit:
Fix player animation

< bug

rubenwiersma commented on src/main/java/com/sem/btrouble/TestState.java:
```diff
@@ -0,0 +1,163 @@
+    @Override
+    public void init(GameContainer gc, StateBasedGame sbg) throws SlickException {
+        // load font from a .ttf file
+        try {
```
Deze code zou in een aparte methode kunnen, nu heeft de init namelijk de extra responsibility van het inladen van een font wat beter naar een andere methode kan worden gedelegeerd.

hungnguyen94 replied:
Ja, dit was een testview voor de MVC. Kan je eigenlijk negeren/verwijderen voor de CollisionHandler strategy pattern.

rubenwiersma commented on src/main/java/com/sem/btrouble/TestState.java:
```diff
@@ -0,0 +1,163 @@
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        drawables = new ArrayList<Drawable>();
```
Waarom heb je drawables eerst als List<Drawable> geinitialiseerd en implementeer je 'm met een ArrayList? Kan je niet direct als ArrayList initialiseren?

hungnguyen94 replied:
List is de interface. ArrayList is een implementatie van de List. Je zou hier drawables ook kunnen initialiseren met een LinkedList oid. Door List te gebruiken, blijft het wat flexibeler.

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Collidable.java:
```diff
@@ -0,0 +1,51 @@
+    float getY();
+
+    /**
+     * Checks for intersection with a shape.
```
intersection with another collidable

hungnguyen94 replied:
Tja, refactor problems.

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/CollisionHandler.java:
```diff
@@ -103,9 +93,9 @@ public int getSize() {
+     * @return - true if Collidable has collided
      */
-    public boolean checkCollision(Shape self) {
+    public boolean checkCollision(Collidable self) {
```
Nice, veel beter dan vorige versie!

rubenwiersma commented on src/main/java/com/sem/btrouble/model/Game.java:
```diff
@@ -0,0 +1,118 @@
+            System.out.println("Level restart");
+        } else {
+            // End game.
+            System.out.println("Game has ended");
```
Hier kunnen we nog een state voor aanmaken die "Game over" toont.

rubenwiersma commented on src/test/java/com/sem/btrouble/CollisionHandlerTest.java:
```diff
@@ -240,12 +240,12 @@ public void checkCollisionTestTrue() {
+        when(player.intersectsCollidable(bubble)).thenReturn(true);
+        when(bubble.intersectsCollidable(player)).thenReturn(true);
         assertTrue(collisionHandler.checkCollision(listCollidables));
     }
```
Zelfs tests!

rubenwiersma commented:
Wat mij betreft is het goed! Nog één iemand reviewen en dan kan 'ie mergen.

Commit:
Add better javadoc to Collidable interface

and it's implementations

< documentation

Commit:
Fire collisionAction of both collided objects on

on collision.
< logic

Commit:
Use a CopyOnWriteArraySet to hold all collidables.

Instead of copying the list before iterating, using a
CopyOnWriteArraySet automatically prevents concurrent modification
exceptions. Happens because the Set can change during iteration.

Commit:
Fix collisionHandler test

< testing, bug, logic

Commit:
Merge branch 'master' into strategyPatternCollisionHandler

Conflicts:
	src/main/java/com/sem/btrouble/controller/CollisionHandler.java
	src/main/java/com/sem/btrouble/model/Bubble.java

Commit:
Small changes to drawing

< user-interface

trombomart commented:
Ziet er goed uit!
